### PR TITLE
Ban DateTime from the codebase

### LINKED ISSUE #47580: Ban `DateTime` from the codebase
By being mutable, `DateTime` is flawed. We should figure out a way to remove it from the codebase and have some CI check that ensures we don't add it back by mistake. I suppose we could configure psalm to do that check for us.

Help welcome!

### PR DESCRIPTION
**Pull request by WebMamba**
| Q             | A
| ------------- | ---
| Branch?       | 6.2
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | yes
| Tickets       | #47580 
| License       | MIT
| Doc PR        | symfony

As discuss in this issue, the purpose of this PR is to remove DateTime from the code base in favor to DateTimeImmutable. I will process it component by component. Feel free to discuss!

### CODE REVIEW

**Comment by stof** [comment]
you should not mark this PR is fixing the issue, as it only does part of the work. We don't want the issue to get closed automatically by github when only that PR is merged. So having a link to the issue is good, but you should not have `Fix` before it.

Note: I edited the PR description to remove it, but please be careful about that for your next PRs

**Review by stof — CHANGES_REQUESTED** [review]
This component is still using `date_create`, which needs to be migrated to `date_create_immutable` instead

**Comment by WebMamba** [comment]
> you should not mark this PR is fixing the issue, as it only does part of the work. We don't want the issue to get closed automatically by github when only that PR is merged. So having a link to the issue is good, but you should not have `Fix` before it.
> 
> Note: I edited the PR description to remove it, but please be careful about that for your next PRs

thanks for for it @stof, I wasn't aware of this. 

> This component is still using `date_create`, which needs to be migrated to `date_create_immutable` instead

Nice catch thanks !

**Review by stof — DISMISSED** [review]

File: src/Symfony/Component/BrowserKit/Cookie.php [hunk 1]
```diff
@@ -202,13 +202,13 @@ private static function parseDate(string $dateValue): ?string
...
         foreach (self::DATE_FORMATS as $dateFormat) {
-            if (false !== $date = \DateTime::createFromFormat($dateFormat, $dateValue, new \DateTimeZone('GMT'))) {
+            if (false !== $date = \DateTimeImmutable::createFromFormat($dateFormat, $dateValue, new \DateTimeZone('GMT'))) {
                 return $date->format('U');
             }
         }
 
         // attempt a fallback for unusual formatting
-        if (false !== $date = date_create($dateValue, new \DateTimeZone('GMT'))) {
+        if (false !== $date = date_create_immutable($dateValue, new \DateTimeZone('GMT'))) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/BrowserKit/Cookie.php:211]: let's replace all similar calls by `new \DateTimeImmutable()` if that's the same- it is right?

Comment (WebMamba) [reply src/Symfony/Component/BrowserKit/Cookie.php:211]: Not really, if the $dateValue is not valid date_create_immutable return false, instead if you use directly the constructor an exception is thrown.

**Comment by stof** [comment]
@WebMamba looks like I miunderstood what you meant by "I'll process component by component", as it seems like you are adding other components in the same PR (which would invalidate my previous approval as it was only for BrowserKit).

If you want us to review it component by component, separate PRs would be a better fit.

**Comment by WebMamba** [comment]
> @WebMamba looks like I miunderstood what you meant by "I'll process component by component", as it seems like you are adding other components in the same PR (which would invalidate my previous approval as it was only for BrowserKit).
> 
> If you want us to review it component by component, separate PRs would be a better fit.

@nicolas-grekas  asked me to process it like this. I do all the easy changes here and if we spot a tricky one we do a dedicated PR

File: src/Symfony/Component/Cache/CacheItem.php [hunk 2]
```diff
@@ -77,7 +77,7 @@ public function expiresAfter(mixed $time): static
         if (null === $time) {
             $this->expiry = null;
         } elseif ($time instanceof \DateInterval) {
-            $this->expiry = microtime(true) + \DateTime::createFromFormat('U', 0)->add($time)->format('U.u');
+            $this->expiry = microtime(true) . \DateTimeImmutable::createFromFormat('U', 0)->add($time)->format('U.u');
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/CacheItem.php]: `+` vs `.`

Comment (WebMamba) [reply src/Symfony/Component/Cache/CacheItem.php]: Hou ! Sorry for it! Bad mistake!

File: src/Symfony/Component/Messenger/Bridge/Doctrine/Tests/Transport/DoctrineIntegrationTest.php [hunk 3]
```diff
@@ -66,9 +66,9 @@ public function testSendWithDelay()
             ->setParameter('body', '{"message": "Hi i am delayed"}')
             ->execute();
 
-        $available_at = new \DateTime($stmt instanceof Result || $stmt instanceof DriverResult ? $stmt->fetchOne() : $stmt->fetchColumn());
+        $available_at = new \DateTimeImmutable($stmt instanceof Result || $stmt instanceof DriverResult ? $stmt->fetchOne() : $stmt->fetchColumn());
 
-        $now = new \DateTime();
+        $now = new \DateTimeImmutable();
         $now->modify('+60 seconds');
```
Comment (stof) [inline src/Symfony/Component/Messenger/Bridge/Doctrine/Tests/Transport/DoctrineIntegrationTest.php]: This is broken. You need to assign the return value when using DateTimeImmutable, as `->modify` does not mutate it.

File: src/Symfony/Component/Messenger/Bridge/Doctrine/Transport/Connection.php [hunk 4]
```diff
@@ -348,7 +348,7 @@ public function getExtraSetupSqlForTable(Table $createdTable): array
 
     private function createAvailableMessagesQueryBuilder(): QueryBuilder
     {
-        $now = new \DateTime();
+        $now = new \DateTimeImmutable();
         $redeliverLimit = (clone $now)->modify(sprintf('-%d seconds', $this->configuration['redeliver_timeout']));
```
Comment (stof) [inline src/Symfony/Component/Messenger/Bridge/Doctrine/Transport/Connection.php]: no need to clone it anymore

File: src/Symfony/Component/Mime/Tests/Header/DateHeaderTest.php [hunk 5]
```diff
@@ -35,7 +35,7 @@ public function testDateTimeCanBeSetBySetter()
 
     public function testDateTimeIsConvertedToImmutable()
     {
-        $dateTime = new \DateTime();
+        $dateTime = new \DateTimeImmutable();
```
Comment (stof) [inline src/Symfony/Component/Mime/Tests/Header/DateHeaderTest.php]: This should be reverted. This test is precisely testing that passing a DateTime as argument gets converted to a DateTimeImmutable.

File: src/Symfony/Component/Mime/Tests/Header/DateHeaderTest.php [hunk 6]
```diff
@@ -44,7 +44,7 @@ public function testDateTimeIsConvertedToImmutable()
 
     public function testDateTimeIsImmutable()
     {
-        $header = new DateHeader('Date', $dateTime = new \DateTime('2000-01-01 12:00:00 Europe/Berlin'));
+        $header = new DateHeader('Date', $dateTime = new \DateTimeImmutable('2000-01-01 12:00:00 Europe/Berlin'));
```
Comment (stof) [inline src/Symfony/Component/Mime/Tests/Header/DateHeaderTest.php]: This defeats the purpose of the test, as `$dateTime->setDate` on next line would not mutate the object passed to the DateHeader anymore. This test is also about covering what happens when passing a mutable DateTime as argument, to ensure we convert it to DateTimeImmutable early enough.

**Review by stof — CHANGES_REQUESTED** [review]

**Comment by WebMamba** [comment]
Thanks, @stof for your review! All the things you spotted are corrected now! 👍

File: src/Symfony/Component/DependencyInjection/Tests/Compiler/ResolveClassPassTest.php [hunk 7]
```diff
@@ -56,7 +56,7 @@ public function provideInvalidClassId()
     {
         yield [\stdClass::class];
         yield ['bar'];
-        yield [\DateTime::class];
```
Comment (OskarStark) [inline src/Symfony/Component/DependencyInjection/Tests/Compiler/ResolveClassPassTest.php:59]: we could keep DateTime here and add DateTimeImmutable, right?

Comment (WebMamba) [reply src/Symfony/Component/DependencyInjection/Tests/Compiler/ResolveClassPassTest.php:59]: Yes we can, but I don't think it adds any value to the test, actually my change here doesn't change any logic. I just made this stay consistent

File: src/Symfony/Component/Notifier/Bridge/Discord/Embeds/DiscordEmbed.php [hunk 8]
```diff
@@ -49,7 +49,7 @@ public function url(string $url): static
     /**
      * @return $this
      */
-    public function timestamp(\DateTime $timestamp): static
+    public function timestamp(\DateTimeInterface $timestamp): static
```
Comment (WebMamba) [inline src/Symfony/Component/Notifier/Bridge/Discord/Embeds/DiscordEmbed.php:52]: I am not sure about this change.

Comment (stof) [reply src/Symfony/Component/Notifier/Bridge/Discord/Embeds/DiscordEmbed.php:52]: this change looks good

File: src/Symfony/Component/VarDumper/Tests/Caster/DateCasterTest.php [hunk 9]
```diff
@@ -30,10 +30,10 @@ class DateCasterTest extends TestCase
      */
     public function testDumpDateTime($time, $timezone, $xDate, $xTimestamp)
     {
-        $date = new \DateTime($time, new \DateTimeZone($timezone));
+        $date = new \DateTimeImmutable($time, new \DateTimeZone($timezone));
```
Comment (stof) [inline src/Symfony/Component/VarDumper/Tests/Caster/DateCasterTest.php:49]: This tests should be preserved. We still want to cover the dumping of a DateTime to make sure we support it. We should have separate tests covering DateTimeImmutable instead.

File: src/Symfony/Component/VarDumper/Tests/Caster/SplCasterTest.php [hunk 10]
```diff
@@ -159,7 +159,7 @@ public function testCastObjectStorageIsntModified()
     public function testCastObjectStorageDumpsInfo()
     {
         $var = new \SplObjectStorage();
-        $var->attach(new \stdClass(), new \DateTime());
+        $var->attach(new \stdClass(), new \DateTimeImmutable());
 
         $this->assertDumpMatchesFormat('%ADateTime%A', $var);
```
Comment (stof) [inline src/Symfony/Component/VarDumper/Tests/Caster/SplCasterTest.php]: should be updated to look for `DateTimeImmutable` and not `DateTime`

File: src/Symfony/Component/VarDumper/Tests/Caster/DateCasterTest.php (hunk 9 above)
Comment (WebMamba) [reply src/Symfony/Component/VarDumper/Tests/Caster/DateCasterTest.php:49]: Ok thanks @stof 😁

File: src/Symfony/Component/VarDumper/Tests/Caster/SplCasterTest.php (hunk 10 above)
Comment (WebMamba) [reply src/Symfony/Component/VarDumper/Tests/Caster/SplCasterTest.php]: Fixed

File: src/Symfony/Component/Console/Logger/ConsoleLogger.php [hunk 11]
```diff
@@ -106,7 +106,7 @@ private function interpolate(string $message, array $context): string
             if (null === $val || \is_scalar($val) || $val instanceof \Stringable) {
                 $replacements["{{$key}}"] = $val;
             } elseif ($val instanceof \DateTimeInterface) {
-                $replacements["{{$key}}"] = $val->format(\DateTime::RFC3339);
+                $replacements["{{$key}}"] = $val->format(\DateTimeImmutable::RFC3339);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Console/Logger/ConsoleLogger.php]: for all consts, we'd better reference the interface all the time

File: src/Symfony/Component/Yaml/Tests/InlineTest.php [hunk 12]
```diff
@@ -573,10 +573,10 @@ public function testParseTimestampAsUnixTimestampByDefault(string $yaml, int $ye
      */
     public function testParseTimestampAsDateTimeObject(string $yaml, int $year, int $month, int $day, int $hour, int $minute, int $second, int $microsecond, string $timezone)
     {
-        $expected = new \DateTime($yaml);
-        $expected->setTimeZone(new \DateTimeZone('UTC'));
-        $expected->setDate($year, $month, $day);
-        $expected->setTime($hour, $minute, $second, $microsecond);
+        $expected = new \DateTimeImmutable($yaml);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Yaml/Tests/InlineTest.php]: let's do that fluently when possible: 
```php
$expected = (new \DateTimeImmutable($yaml))
    ->setTimezone(new \DateTimeZone('UTC'))
...
```

File: src/Symfony/Component/Console/Logger/ConsoleLogger.php (hunk 11 above)
Comment (WebMamba) [reply src/Symfony/Component/Console/Logger/ConsoleLogger.php]: Thanks for your review, It should be fixed with my last commit

File: src/Symfony/Component/Yaml/Tests/InlineTest.php (hunk 12 above)
Comment (WebMamba) [reply src/Symfony/Component/Yaml/Tests/InlineTest.php]: should be fixed

File: src/Symfony/Component/VarDumper/Caster/ResourceCaster.php [hunk 13]
```diff
@@ -77,7 +77,7 @@ public static function castOpensslX509($h, array $a, Stub $stub, bool $isNested)
         $a += [
             'subject' => new EnumStub(array_intersect_key($info['subject'], ['organizationName' => true, 'commonName' => true])),
             'issuer' => new EnumStub(array_intersect_key($info['issuer'], ['organizationName' => true, 'commonName' => true])),
-            'expiry' => new ConstStub(date(\DateTime::ISO8601, $info['validTo_time_t']), $info['validTo_time_t']),
+            'expiry' => new ConstStub(date(\DateTimeImmutable::ISO8601, $info['validTo_time_t']), $info['validTo_time_t']),
```
Comment (nicolas-grekas) [inline src/Symfony/Component/VarDumper/Caster/ResourceCaster.php]: DateTimeInterface

**Review by nicolas-grekas — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @WebMamba.

**Pull request merged**